### PR TITLE
Interleaved TCP mode at RTSP source side.

### PR DESCRIPTION
**Pull request by kzkysdjpn**
This is RTSP TCP interleaved mode at source server side.

### CODE REVIEW

File: lib/RTSP/Server/Listener.pm [hunk 1]
```diff
@@ -117,6 +117,23 @@ sub listen {
                 $cleanup->();
             },
             on_read => sub {
+		my	$len;
+		my	$magic;
+		$magic = unpack("C", $handle->{rbuf});
+		if($magic == 0x24){
```
Comment (revmischa) [inline lib/RTSP/Server/Listener.pm]: comment what this magic number is?

File: lib/RTSP/Server/Listener.pm [hunk 2]
```diff
@@ -117,6 +117,23 @@ sub listen {
...
+			$len = unpack("n", substr($handle->{rbuf}, 2, 4));
+			$len += 4;
+			$handle->push_read(
+				chunk => $len, sub{
+                                        my (undef, $data) = @_;
+                                        my $chan;
+                                        $chan = unpack("C", substr($data, 1, 1));
+                                        $conn->write_interleaved_rtp($chan, substr($data, 4));
```
Comment (revmischa) [inline lib/RTSP/Server/Listener.pm]: make sure `use bytes` is in effect so this isn't string-based

File: lib/RTSP/Server/Source/Connection.pm [hunk 3]
```diff
@@ -170,28 +190,87 @@ sub setup {
...
+        $transport =~ m/interleaved=(\d+)(?:\-(\d+))/smi;
+    if(index($transport, "interleaved=") != -1){
+        unless(length($chan_str[0])){
+            return $self->bad_request;
+        }
+        unless(length($chan_str[1])){
+            return $self->bad_request;
+        }
+        $chan[0] = $chan_str[0] + 0;
+        $chan[1] = $chan_str[1] + 0;
```
Comment (revmischa) [inline lib/RTSP/Server/Source/Connection.pm]: may want to assert these aren't equal?

File: lib/RTSP/Server/Source/Connection.pm [hunk 4]
```diff
@@ -170,28 +190,87 @@ sub setup {
...
+        }else{
+            $dest = sockaddr_in6($stream->rtp_start_port, Socket6::inet_pton(AF_INET6, "localhost"));
+        }
+        unless (connect $sock, $dest){
+            return $self->bad_request;
+        }
+        $self->channel_sockets->{$chan[0] . ""} = $sock;
+
+        # create UDP socket for internal RTCP packet stream
+        socket my($sock_rtcp), $self->addr_family, SOCK_DGRAM, $udp_proto;
```
Comment (revmischa) [inline lib/RTSP/Server/Source/Connection.pm]: what is being done with the RTCP stream?

File: rtsp-server.pl [hunk 5]
```diff
@@ -10,6 +10,23 @@
 # you may pass your own options in here or via command-line
 my $srv = RTSP::Server->new_with_options(
 );
+$srv->add_source_update_callback(sub {
+	print("Add source operation\n");
```
Comment (revmischa) [inline rtsp-server.pl]: funny indentation here

**Comment by revmischa** [comment]
@kzkysdjpn sup

**Comment by kzkysdjpn** [comment]
I am sorry.
This pull request has bugs.
I will fixed and request again, later.

**Comment by revmischa** [comment]
Woah man just do a push to fix the things above! No need to close the PR

**Pull request closed without merging**